pkg/controller/jobframework/reconciler.go

```diff
@@ -596,8 +596,8 @@ func (r *JobReconciler) ensureOneWorkload(ctx context.Context, job GenericJob, o
 		toDelete = toDelete[1:]
 	}
 
-	// If there is no matching workload and the job is running, suspend it.
-	if match == nil && !job.IsSuspended() {
+	// If there is no matching workload suspend job.
+	if match == nil {
```

Comment:
What happens in the case where there is a suspended Job without restored nodeSelector which means the reconciler failed to restore nodeSelector in the previous reconciliation.

I'm suspecting that the Worlload is accidentally removed, right?
Could you add a unit test in the batch/job controller for the assumed situations?

Reply:
You mean this case?

https://github.com/kubernetes-sigs/kueue/blob/dfa182907281dea39fd227aea58abf1374727bef/pkg/controller/jobs/job/job_controller_test.go#L2552-L2601

Reply:
Looks like removing workloads happening after stop, so it's nothing change. 

https://github.com/kubernetes-sigs/kueue/blob/dfa182907281dea39fd227aea58abf1374727bef/pkg/controller/jobframework/reconciler.go#L600-L641

Reply:
Looks like this solution will not work on another cases. 

For example here:

https://github.com/kubernetes-sigs/kueue/blob/dfa182907281dea39fd227aea58abf1374727bef/pkg/controller/jobframework/reconciler.go#L484-L492

Reply:
> Looks like this solution will not work on another cases.

How does the code change affect the code you reference here?

Reply:
Maybe it's a good idea to create something like transactions using labels. We can add some label when we start suspending job and delete this label on the last update.

```go
func (j *Job) Stop(ctx context.Context, c client.Client, podSetsInfo []podset.PodSetInfo, _ jobframework.StopReason, eventMsg string) (bool, error) {
	stoppedNow := false
	if !j.IsSuspended() {
		j.Suspend()
		j.ObjectMeta.Labels[StoppingLabel] = "true"
		if err := c.Update(ctx, j.Object()); err != nil {
			return false, fmt.Errorf("suspend: %w", err)
		}
		stoppedNow = true
	}

	// Reset start time if necessary, so we can update the scheduling directives.
	if j.Status.StartTime != nil {
		j.Status.StartTime = nil
		if err := c.Status().Update(ctx, j.Object()); err != nil {
			return stoppedNow, fmt.Errorf("reset status: %w", err)
		}
	}

	j.RestorePodSetsInfo(podSetsInfo)
	delete(j.ObjectMeta.Labels, StoppingLabel)
	if err := c.Update(ctx, j.Object()); err != nil {
		return false, fmt.Errorf("restore info: %w", err)
	}
	return stoppedNow, nil
}

...

func (j *Job) IsSuspended() bool {
    return j.Spec.Suspend != nil && *j.Spec.Suspend && j.Labels[StoppingLabel] != "true"
}

```
 
WDYT @alculquicondor @mimowo

Reply:
Using labels seems awkward, if anything we could use conditions for this purpose, but I'm not clear if this is needed.

Reply:
> > Looks like this solution will not work on another cases.
> 
> How does the code change affect the code you reference here? Still, I think we should really test this e2e.

I mean that it will not "re-stop" job because we have:

https://github.com/kubernetes-sigs/kueue/blob/dfa182907281dea39fd227aea58abf1374727bef/pkg/controller/jobframework/reconciler.go#L445-L446

Reply:
> Using labels seems awkward, if anything we could use conditions for this purpose, but I'm not clear if this is needed.

Agree, but on this case we will know for sure that it stopped.

Reply:
I see. In that case it seems trickier than initially thought.